File: js_solutions/JZ20minStack.java
import java.util.Stack;

public class Solution {
    private Stack<Integer> stack = new Stack<>();
    private Stack<Integer> minStack = new Stack<>();
    public void push(int node) {
        int minValue;
        if(minStack.isEmpty()) {
            minValue = node;
        } else {
            minValue = Math.min(minStack.peek(), node);
        }
        stack.push(node);
        minStack.push(minValue);
    }
    
    public void pop() {
        if(!stack.isEmpty()) {
            stack.pop();
            minStack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int min() {
        return minStack.peek();
    }
}

File: js_solutions/JZ20minStack.js
// 定义栈的数据结构，请在该类型中实现一个能够得到栈中所含最小元素的min函数（时间复杂度应为O（1））。


const minStack = [], stack  = [];
let minValue = null;

function push(node)
{
    // write code here
    if(minValue) {
        minValue = Math.min(node, minValue)
    } else {
        minValue = node
    }
    stack.push(node)
    minStack.push(minValue)
}
function pop()
{
    // write code here
    const temp = stack.pop();
    minStack.pop();
    return temp;
}
function top()
{
    // write code here
    return stack[stack.length - 1]
}
function min()
{
    // write code here
    return minStack[minStack.length - 1]
}